File: bot.js
#!/usr/bin/env node
// Локальный режим для разработки: long polling вместо вебхука.
// Тонкая обёртка — логика в lib/handle.js, ровно та же, что в проде.
//
//   node --env-file=.env bot.js
//
// В проде полинг невозможен: serverless-функция существует только во время
// запроса, а без запроса некому крутить цикл.
//
// Отличие от прода одно: здесь мы дожидаемся handleUpdate, а на Vercel он
// уезжает в waitUntil. Ядро об этом не знает.

import { call, TelegramError } from "./lib/telegram.js";
import { handleUpdate } from "./lib/handle.js";

let offset;
let running = true;

process.on("SIGINT", () => {
  running = false;
  console.log("\nОстанавливаюсь.");
  process.exit(0);
});

console.log("Слушаю. Ctrl-C чтобы остановить.");

while (running) {
  try {
    const updates = await call(
      "getUpdates",
      { timeout: 30, offset, allowed_updates: ["message"] },
      { timeoutMs: 40_000 },
    );

    for (const update of updates) {
      offset = update.update_id + 1;
      await handleUpdate(update);
    }
  } catch (error) {
    if (error instanceof TelegramError || error.name === "TimeoutError") {
      console.warn(`[polling] ${error.message}, повтор через 3 с`);
      await new Promise((resolve) => setTimeout(resolve, 3000));
    } else {
      throw error;
    }
  }
}

File: bot.rb
#!/usr/bin/env ruby
# Локальный режим для разработки (Этап 1): long polling вместо вебхука.
# Тонкая обёртка — вся логика в lib/handle.rb, ровно та же, что в проде.
#
#   ruby bot.rb
#
# В проде полинг невозможен: serverless-функция существует только во время
# запроса, а без запроса некому крутить цикл.

require_relative "lib/env"
require_relative "lib/handle"

Signal.trap("INT") { exit(0) }

offset = nil
puts "Слушаю. Ctrl-C чтобы остановить."

loop do
  updates = Telegram.call(
    "getUpdates",
    { timeout: 30, offset: offset, allowed_updates: ["message"] }.compact,
    read_timeout: 40
  )

  updates.each do |update|
    offset = update["update_id"] + 1
    Handle.handle_update(update)
  end
rescue Telegram::Error, Net::ReadTimeout, Net::OpenTimeout => e
  warn "[polling] #{e.message}, повтор через 3 с"
  sleep 3
end
